### PR TITLE
Strict type `Docker::Requirement` and `Docker::Tag`

### PR DESCRIPTION
**Pull request by raj-meka**

### CODE REVIEW

File: docker/lib/dependabot/docker/requirement.rb [hunk 1]
```diff
@@ -11,21 +11,24 @@ module Docker
     # Lifted from the bundler package manager
     class Requirement < Dependabot::Requirement
       extend T::Sig
+      extend T::Helpers
```
Comment (JamieMagee) [inline docker/lib/dependabot/docker/requirement.rb:14]: I'm not sure you need this `extend`? What feature of Sorbet do you need from it?

File: docker/lib/dependabot/docker/tag.rb [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
 # frozen_string_literal: true
 
 require "dependabot/docker/file_parser"
+require "sorbet-runtime"
 
 module Dependabot
   module Docker
     class Tag
+      extend T::Sig
+      extend T::Helpers
```
Comment (JamieMagee) [inline docker/lib/dependabot/docker/tag.rb:11]: Same here as above. What feature of Sorbet requires `T::Helpers`?

File: docker/lib/dependabot/docker/tag.rb [hunk 3]
```diff
@@ -17,25 +21,30 @@ class Tag
...
 
+      sig { returns(T::Boolean) }
       def digest?
         name.match?(FileParser::DIGEST)
       end
 
+      sig { returns(T::Boolean) }
       def looks_like_prerelease?
-        numeric_version.match?(/[a-zA-Z]/)
+        T.must(numeric_version).match?(/[a-zA-Z]/)
```
Comment (JamieMagee) [inline docker/lib/dependabot/docker/tag.rb:44]: Does this work?

```suggestion
        numeric_version&.match?(/[a-zA-Z]/)
```

Generally, we want to use Ruby's [safe navigation operator (`&.`)][1] where possible, and `T.must` only if safe navigation isn't possible. The reason being `T.must` can throw, when the parameter is `nil`.

[1]: https://thoughtbot.com/blog/ruby-safe-navigation

File: docker/lib/dependabot/docker/tag.rb [hunk 4]
```diff
@@ -51,47 +60,55 @@ def comparable_to?(other)
...
       end
 
+      sig { returns(T::Boolean) }
       def canonical?
         return false unless numeric_version
         return true if name == numeric_version
 
         # .NET tags are suffixed with -sdk
-        return true if name == numeric_version + "-sdk"
+        return true if name == T.must(numeric_version) + "-sdk"
```
Comment (JamieMagee) [inline docker/lib/dependabot/docker/tag.rb:86]: Does this work?

```suggestion
        return true if numberic_version && name == numeric_version + "-sdk"
```

File: docker/lib/dependabot/docker/tag.rb [hunk 5]
```diff
@@ -51,47 +60,55 @@ def comparable_to?(other)
...
 
-        name == "jdk-" + numeric_version
+        name == "jdk-" + T.must(numeric_version)
```
Comment (JamieMagee) [inline docker/lib/dependabot/docker/tag.rb:88]: Similar to above
```suggestion
        numeric_version && name == "jdk-" + numeric_version
```

If not, keep it as is.

File: docker/lib/dependabot/docker/tag.rb [hunk 6]
```diff
@@ -51,47 +60,55 @@ def comparable_to?(other)
...
       end
 
+      sig { returns T.nilable(String) }
       def prefix
-        name.match(NAME_WITH_VERSION).named_captures.fetch("prefix")
+        T.must(name.match(NAME_WITH_VERSION)).named_captures.fetch("prefix")
```
Comment (JamieMagee) [inline docker/lib/dependabot/docker/tag.rb:93]:
```suggestion
        name.match(NAME_WITH_VERSION)&.named_captures.fetch("prefix")
```

File: docker/lib/dependabot/docker/tag.rb [hunk 7]
```diff
@@ -51,47 +60,55 @@ def comparable_to?(other)
...
       end
 
+      sig { returns T.nilable(String) }
       def suffix
-        name.match(NAME_WITH_VERSION).named_captures.fetch("suffix")
+        T.must(name.match(NAME_WITH_VERSION)).named_captures.fetch("suffix")
```
Comment (JamieMagee) [inline docker/lib/dependabot/docker/tag.rb:98]:
```suggestion
        name.match(NAME_WITH_VERSION)&.named_captures.fetch("suffix")
```

File: docker/lib/dependabot/docker/tag.rb [hunk 8]
```diff
@@ -51,47 +60,55 @@ def comparable_to?(other)
...
       end
 
+      sig { returns T.nilable(String) }
       def version
-        name.match(NAME_WITH_VERSION).named_captures.fetch("version")
+        T.must(name.match(NAME_WITH_VERSION)).named_captures.fetch("version")
```
Comment (JamieMagee) [inline docker/lib/dependabot/docker/tag.rb:103]:
```suggestion
        name.match(NAME_WITH_VERSION)&.named_captures.fetch("version")
```

File: docker/lib/dependabot/docker/tag.rb [hunk 9]
```diff
@@ -51,47 +60,55 @@ def comparable_to?(other)
...
 
+      sig { returns(Symbol) }
       def format
         return :sha_suffixed if name.match?(/(^|\-g?)[0-9a-f]{7,}$/)
-        return :year_month if version.match?(/^[12]\d{3}(?:[.\-]|$)/)
-        return :year_month_day if version.match?(/^[12](?:\d{5}|\d{7})(?:[.\-]|$)/)
-        return :build_num if version.match?(/^\d+$/)
+        return :year_month if T.must(version).match?(/^[12]\d{3}(?:[.\-]|$)/)
+        return :year_month_day if T.must(version).match?(/^[12](?:\d{5}|\d{7})(?:[.\-]|$)/)
+        return :build_num if T.must(version).match?(/^\d+$/)
```
Comment (JamieMagee) [inline docker/lib/dependabot/docker/tag.rb:111]:
```suggestion
        return :year_month if version&.match?(/^[12]\d{3}(?:[.\-]|$)/)
        return :year_month_day if version&.match?(/^[12](?:\d{5}|\d{7})(?:[.\-]|$)/)
        return :build_num if version&.match?(/^\d+$/)
```

File: docker/lib/dependabot/docker/tag.rb [hunk 10]
```diff
@@ -100,25 +117,28 @@ def format
         # That means only "22-ea-7" will be considered as a viable update
         # candidate for "21-ea-32", since it's the only one that respects that
         # format.
-        if version.match?(WORDS_WITH_BUILD)
-          return :"<version>#{version.match(WORDS_WITH_BUILD).to_s.gsub(/-[0-9]+/, '-<build_num>')}"
+        if T.must(version).match?(WORDS_WITH_BUILD)
```
Comment (JamieMagee) [inline docker/lib/dependabot/docker/tag.rb:120]:
```suggestion
        if version&.match?(WORDS_WITH_BUILD)
```

File: docker/lib/dependabot/docker/tag.rb [hunk 11]
```diff
@@ -100,25 +117,28 @@ def format
...
+          return :"<version>#{T.must(version).match(WORDS_WITH_BUILD).to_s.gsub(/-[0-9]+/, '-<build_num>')}"
```
Comment (JamieMagee) [inline docker/lib/dependabot/docker/tag.rb:121]: IMO this `T.must` is okay, as we already verified on the line above that it's not `nil`.

File: docker/lib/dependabot/docker/tag.rb [hunk 12]
```diff
@@ -100,25 +117,28 @@ def format
...
 
         :normal
       end
 
+      sig { returns(T.nilable(String)) }
       def numeric_version
         return unless comparable?
 
-        version.gsub(/kb/i, "").gsub(/-[a-z]+/, "").downcase
+        T.must(version).gsub(/kb/i, "").gsub(/-[a-z]+/, "").downcase
```
Comment (JamieMagee) [inline docker/lib/dependabot/docker/tag.rb:131]:
```suggestion
        version&.gsub(/kb/i, "").gsub(/-[a-z]+/, "").downcase
```

**Review by JamieMagee — CHANGES_REQUESTED** [review]
Needs some clarification around use of `T.must`. I understand that this is the autocorrection that Sorbet suggests in your IDE, but before you accept it, double check if safe navigation (`&.`) would work first.

**Review by JamieMagee — APPROVED** [review]

**Review by abdulapopoola — APPROVED** [review]

**Pull request merged**